### PR TITLE
FIX '[Error 5] Access is denied' under Windows

### LINKED ISSUE #354: Joblib import error after a run of failed gridsearch cv in Azure VM
Hi , I am running gridsearchcv on a 16 core Azure VM. While the same code works fine on my PC, I am getting the following error on the VM. It is fitting all the folds as I can see from verbose(=10) but it is throwing up this error after finishing all the folds. This is occurring after fitting all CV-folds but before calculating the "best_score_" and "best_estimator_".Below is the traceback. (I am getting the same error for n_jobs=-1 or a number less than 16 and refit is True)

```
CV] ml__C=0.5, ml__penalty=l1, ml__tol=0.01, ml__max_iter=300, ml__warm_start=False 
[CV] ml__C=0.5, ml__penalty=l1, ml__tol=0.01, ml__max_iter=300, ml__warm_start=False, score=0.620114 - 47.6s
Traceback (most recent call last):
File "", line 1, in 
File "C:\Anaconda2\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 699, in runfile
execfile(filename, namespace)
File "C:\Anaconda2\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 74, in execfile
exec(compile(scripttext, filename, 'exec'), glob, loc)
File "C:/Users/Preston/Documents/Python Scripts/ml_cpt_cv.py", line 111, in 
model.fit(x_train, y_train)
File "C:\Anaconda2\lib\site-packages\sklearn\grid_search.py", line 804, in fit
return self.fit(X, y, ParameterGrid(self.param_grid))
File "C:\Anaconda2\lib\site-packages\sklearn\grid_search.py", line 553, in fit
for parameters in parameter_iterable
File "C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\parallel.py", line 818, in __call
self.terminate_pool()
File "C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\parallel.py", line 549, in terminate_pool
self._pool.terminate() # terminate does a join()
File "C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\pool.py", line 583, in terminate
super(MemmapingPool, self).terminate()
File "C:\Anaconda2\lib\multiprocessing\pool.py", line 465, in terminate
self._terminate()
File "C:\Anaconda2\lib\multiprocessing\util.py", line 207, in __call
res = self._callback(self._args, *self._kwargs)
File "C:\Anaconda2\lib\multiprocessing\pool.py", line 513, in _terminate_pool
p.terminate()
File "C:\Anaconda2\lib\multiprocessing\process.py", line 137, in terminate
self._popen.terminate()
File "C:\Anaconda2\lib\multiprocessing\forking.py", line 312, in terminate
_subprocess.TerminateProcess(int(self._handle), TERMINATE)
WindowsError: [Error 5] Access is denied
```

ALSO, in the very next run of the code, I am getting the following Joblib error.
Fitting 4 folds for each of 3 candidates, totalling 12 fits

```
Traceback (most recent call last):
File "", line 1, in 
File "C:\Anaconda2\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 699, in runfile
execfile(filename, namespace)
File "C:\Anaconda2\lib\site-packages\spyderlib\widgets\externalshell\sitecustomize.py", line 74, in execfile
exec(compile(scripttext, filename, 'exec'), glob, loc)
File "C:/Users/Preston/Documents/Python Scripts/ml_cpt_cv.py", line 111, in 
model.fit(x_train, y_train)
File "C:\Anaconda2\lib\site-packages\sklearn\grid_search.py", line 804, in fit
return self.fit(X, y, ParameterGrid(self.param_grid))
File "C:\Anaconda2\lib\site-packages\sklearn\grid_search.py", line 553, in fit
for parameters in parameter_iterable
File "C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\parallel.py", line 766, in __call
n_jobs = self.initialize_pool()
File "C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\parallel.py", line 515, in initialize_pool
raise ImportError('[joblib] Attempting to do parallel computing '
ImportError: [joblib] Attempting to do parallel computing without protecting your import on a system that does not support forking. To use parallel-computing in a script, you must protect your main loop using "if __name == 'main'". Please see the joblib documentation on Parallel for more information.
```

This is the code,

```python
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.svm import LinearSVC
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import f1_score, precision_score, recall_score, precision_recall_curve, accuracy_score,confusion_matrix
import sklearn.grid_search
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn import pipeline,metrics, grid_search

if name=="main":
train_data=pd.read_csv("train.csv",delimiter=",")
test_data=pd.read_csv("test.csv",delimiter=",")
n_cv=int(raw_input("Please tell us how many fold cross validation you would like to perform,please give a value between 3-10"))
assert (n_cv>=3) and (n_cv<=10), "Looks like the number of folds is not in between 3 and 10"
#scl = StandardScaler()
y_train=train_data["class"]
y_test=test_data["class"]
x_train=train_data
x_test=test_data
#x_test=list(x_test)
y_train=list(y_train)
y_train=[1 if i==True else 0 for i in y_train]

y_test=list(y_test)
y_test=[1 if i==True else 0 for i in y_test]

type_of_ml_model=raw_input("Please input rf for random forests, lr for logistic regression, svm for support vector machines and gbdt for gradient boosting")
if type_of_ml_model=="rf":
ml_model=RandomForestClassifier(random_state=5)
clf = pipeline.Pipeline([('ml', ml_model)])
param_grid = {
'ml__n_estimators':[100,500],
'ml__max_features':["auto",None,"log2"],
'ml__max_depth':[5,4,3],
'ml__min_samples_split':[1,2],
'ml__oob_score':[True,False],
'ml__class_weight':["balanced","balanced_subsample"]
}

elif type_of_ml_model=="lr":
ml_model=LogisticRegression(class_weight="balanced",random_state=5)
clf = pipeline.Pipeline([('ml', ml_model)])
param_grid = {#'ml__fit_intercept':[True],'ml__intercept_scaling':[1,2],
'ml__tol':[0.01],'ml__max_iter':[300],'ml__warm_start':[False],
'ml__C': [0.5,0.2,0.6],
'ml__penalty':["l1"]}

elif type_of_ml_model=="svm":
ml_model=LinearSVC(class_weight="balanced",random_state=5)
clf = pipeline.Pipeline([('ml', ml_model)])
param_grid = {'ml__loss':["squared_hinge"],'ml__dual':[False],'ml__fit_intercept':[True],
'ml__intercept_scaling':[2,3],
'ml__tol':[0.001,0.0001],'ml__max_iter':[200],
'ml__C': [1,0.8],
'ml__penalty':["l1"]}

elif type_of_ml_model=="gbdt":
ml_model=GradientBoostingClassifier(random_state=5)
clf = pipeline.Pipeline([('ml', ml_model)])
param_grid = {'ml__loss':["deviance"],
'ml__n_estimators':[100,200],
'ml__max_features':["auto",None],
'ml__max_depth':[4,3],
'ml__min_samples_split':[1,2],
'ml__subsample':[1.0],'ml__warm_start':[False],
'ml__min_samples_leaf':[1,2]
}

precision_scorer = metrics.make_scorer(precision_score, greater_is_better = True)

# Initialize Grid Search Model
model = grid_search.GridSearchCV(estimator = clf, param_grid=param_grid, scoring=precision_scorer,
verbose=10,n_jobs=-1, iid=True, refit=True, cv=n_cv)

# Fit Grid Search Model
model.fit(x_train, y_train)
```

Is there something in parameters that we need to change when we are running on a VM?

### PR DESCRIPTION
**Pull request by ogrisel**
This should fix #354.

### CODE REVIEW

File: joblib/pool.py [hunk 1]
```diff
@@ -596,5 +597,16 @@ def _cleanup():
...
+        n_retries = 10
+        for i in range(n_retries):
+            try:
+                super(MemmapingPool, self).terminate()
+                break
+            except WindowsError as e:
+                # Workaround  occasional "[Error 5] Access is denied" issue
+                # when trying to terminate a process under windows.
+                sleep(0.1)
+                if i + 1 == n_retries:
```
Comment (lesteve) [inline joblib/pool.py]: Can you not use a else clause if the for loop for this?

**Comment by lesteve** [comment]
Should we try to get @krishnateja614 to test this? Maybe it is too much of a pain since it is happening on Azure ...

**Comment by krishnateja614** [comment]
Guys. I haven't had any problems after changing the timeout. If you want to test something out, I'll have access to the VM in the morning and can do so.

**Comment by lesteve** [comment]
> Guys. I haven't had any problems after changing the timeout.

Patching the python stdlib is not really the nicest of solution though ;-).

If you could test this change and look whether you never get any warning, that would be great.

**Comment by ogrisel** [comment]
@krishnateja614 I would really appreciate it if you could undo the change you did to the standard library and instead test if this change in joblib can fix the crash you observe on your azure configuration.

**Comment by krishnateja614** [comment]
I'll test this once I get access to the VM. So basically,
 reset the timeout to 0.1 in multiprocessing/forking.py  
install joblib from the github repo, then run the code right?

**Comment by krishnateja614** [comment]
@ogrisel , I got the same error after trying with the updated joblib code. Also the joblib import error occurs only after this multiprocessing\forking.py terminateprocess function fails. This update on Joblib doesn't change that right or is there anything else I should be doing?. Again, the access denied error is not occurring for all the values of CV.

**Comment by lesteve** [comment]
The simplest is if you go there: https://raw.githubusercontent.com/ogrisel/joblib/71cc0d96967f5090894d7ae3347e04d0d60281c0/joblib/pool.py save it as a file (Control-S on most browsers) and then copy this file into C:\Anaconda2\lib\site-packages\sklearn\externals\joblib\pool.py.

**Comment by krishnateja614** [comment]
Got it.Trying it out. Will update this comment.

**Comment by ogrisel** [comment]
Interesting that you got the warning: it means that you still have some processes that are slow to die in the background. But at least it does not seem to hurt your program anymore. Will merge.

Thanks for testing the fix in your runtime environment!

**Comment by krishnateja614** [comment]
I don't know whether I should reopen this or not, but I have again ran into the same error as #354, the only difference is the dimensionality of the dataset. The error previously looked corrected when run on dataset of size 8,00,000 * 35 (I don't remember the number of columns  I had exactly, but it's definitely less than 35) but now, I ran into the same error two days ago when running the model on 8,00,000 * 210 dataset. VM is same as before i.e Windows server 12 with 16 cores CPU & 116 GB RAM (AZURE D14_V2).

**Pull request merged**